### ISSUE TITLE
Setting custom stoploss according to base line

### ISSUE THREAD

**Issue by deception83** [freqtrade/freqtrade#5505]
Hi, I'd like to know a few things. Doesoes custom stoploss get set only once per buy per pair? If not how should I make sure it only set once? 

Lastly, I am using ichimoku to find the base line. So with that baseline I want to set the stoploss at my according to the price I bought at percentage distance to the base line at that point of time which means my stop loss should not change after.

Do I have to query the SQLite database for the info? Or are they present at the dataframe? 

Thanks.

**Comment by xmatthias** [comment]
custom stoploss is called once per iteration (by default, roughly every 5 seconds once a trade is open).

If you only want to set the stop once (and not have it trailing) - you'll need to make sure to either only return a price only once - or always return the same price. 

There's several examples in the [advanced documentation](https://www.freqtrade.io/en/stable/strategy-advanced/#custom-stoploss-using-an-indicator-from-dataframe-example) - but best read through the whole page, as it includes [other relevant](https://www.freqtrade.io/en/stable/strategy-advanced/#dataframe-access) tipps for your requirement as well.

I don't see what you're looking for in the database - the trade object passed into the function has the same information than the database - but i don't really see that relevant for your question.

**Comment by deception83** [comment]
So am I right to assume that this function will be called only when an actual buy action has happened and will be call at every iteration until it's has been sold? 

Also, does current_price reflect the current price in the dataframe or at the price it's bought? I assume it's not, so I will have to use profit and current to work out the price that's it's bought. 

I'll take a look at the 2 links you mention and try to work something out. Thanks.

**Comment by xmatthias** [comment]
yes, it's called for the duration of a trade only (no point in calling it when no trade is open ...).

`current_price` corresponds to the current price (based on the configured logic for selling trades). 
`current_profit` is calculated based on that same price as well.

Otherwise it'd be called "open price" ... which is however available through the trade object.

**Comment by deception83** [comment]
```
def custom_stoploss(self, pair: str, trade: 'Trade', current_time: datetime, current_rate: float, current_profit: float, **kwargs) -> float:        
        #Start at next second candle to let prices stabilise
        if current_time - timedelta(minutes=60) > trade.open_date_utc:
        
            # Obtain pair dataframe.
            dataframe, _ = self.dp.get_analyzed_dataframe(pair, self.timeframe)
            
            # In dry/live runs trade open date will not match candle open date therefore it must be 
            # rounded.
            trade_date = timeframe_to_prev_date(self.timeframe, trade.open_date_utc)
            trade_candle = dataframe.loc[dataframe['date'] == trade_date]
            
            if not trade_candle.empty:
                base_line = trade_candle['kijun_sen'].iloc[0]
                open_price = trade_candle['open'].iloc[0]
                initial_stoploss_dist = (open_price / base_line) - 1
                
                #Only change stoploss if there was profit else set initial stoploss
                if current_profit > open_price:
                    current_stoploss_dist = (current_rate / open_price) - 1
                    new_stoploss = (initial_stoploss_dist + current_stoploss_dist)
                    
                else:
                    new_stoploss =  initial_stoploss_dist
                
                if new_stoploss != 0:            
                    # print (new_stoploss)              
                    return new_stoploss        
        return 1

```
I did a backtest with this not sure if its correct or the stoploss became too conservative that my profits took a dive. If its the latter I guess I shall think of a new stoploss plan. =)

**Comment by xmatthias** [comment]
I think you'll need to be careful with using it as is ...

`kijun_sen` is not guaranteed to be below the candle price - so this will strongly depend on your entry signal - and might result in VERY tight stoplosses.

**Comment by deception83** [comment]
I guess I found what works for me. So here goes...

First at `populate_indicators` , I added another set of data while will look at the previous 20 to look for the lowest point and set it as my baseline. 
```
dataframe['lowest'] = dataframe['low'].rolling(window=20).min()
```

Then since I have that data in the trade candle I can just simply reference to that and keep my stoploss price static throughout the whole trade.

```
    def custom_stoploss(self, pair: str, trade: 'Trade', current_time: datetime,
                        current_rate: float, current_profit: float, **kwargs) -> float:
                
        # Obtain pair dataframe.
        dataframe, _ = self.dp.get_analyzed_dataframe(pair, self.timeframe)
        trade_date = timeframe_to_prev_date(self.timeframe, trade.open_date_utc)
        # Look up trade candle.
        trade_candle = dataframe.loc[dataframe['date'] == trade_date]
        
        if not trade_candle.empty:
        
            if current_profit > 0.02: #Let prices stabilize before setting
                base_line = trade_candle['lowest'].iloc[0]
                open_price = trade_candle['open'].iloc[0]
                set_stoploss = (open_price / base_line) - 1
                
                return set_stoploss
            
        return 1
```

Thanks @xmatthias this small-ish adventure brought me into the world of Pandas. So with this I am closing this.
